### PR TITLE
Estimate SedPhos And Recession Coefficient From Rasters

### PR DESCRIPTION
**Pull request by arottersman**
## Overview

- Use the `soilpallland2` raster in the datahub to estimate the mapshed `SedPhos`/`SoilP` value
- Use the `bfi48grd` raster in the datahub to estimate the mapshed `RecessionCoef` value

Connects #1752
Connects #1756  

### Demo
Resulting GMS file

[panther_skin_creek.gms.txt](https://github.com/WikiWatershed/model-my-watershed/files/1311837/panther_skin_creek.gms.txt)

Line 2, pos 2 is not `0.06`
```
1.4.0,0.01,0,10,0,0,0.12504566328764893,13.825475584291853,62287,24254.230851633638,38032.769148366366,1914.431863640184,0,8.460119315139668,0.021151433681949956,28,1963,1990,0.0004415567779477565,7404.395733781011,0,0,0,0.3584968896652178
```

Line 37
```
701.52,816.6336633663367,0.7774450661770334,0.01793083008186061,0.25,0.25
```

## Testing Instructions

- Check out branch
- Restart Celery `vagrant ssh worker -c 'sudo service celeryd restart'`
- Model a WkAoI that isn't cached on the branch and on staging
- Compare line 2, pos 2 and line 37 pos 2 across staging and this branch. They should be different now.

### CODE REVIEW

**Comment by rajadain** [comment]
@azavea-bot rebuild

**Comment by rajadain** [comment]
Taking a look now.

File: src/mmw/apps/modeling/mapshed/tasks.py [hunk 1]
```diff
@@ -319,6 +318,43 @@ def avg_awc(result):
     }
 
 
+
+@shared_task(throws=Exception)
```
Comment (rajadain) [inline src/mmw/apps/modeling/mapshed/tasks.py]: This task has too many blank lines above it, causing the build to be marked as unstable.

**Comment by mmcfarland** [comment]
@ajrobbins Can you share this GMS line with BME to make sure the value range is within what he expects?

**Comment by ajrobbins** [comment]
Yep!

**Comment by rajadain** [comment]
So we're not using the Soil Texture raster at all with this?

**Comment by arottersman** [comment]
> So we're not using the Soil Texture raster at all with this?

I'll wait and see the result of slack discussion, but ready to add another commit to remove it as a layer from that job...didn't realized it wasn't getting used by anything else.

**Comment by rajadain** [comment]
Do keep this in a separate commit, but this patch removes `nlcd_soils`, and renames `nlcd_soil_census` to `nlcd_soil` (which was initially named `_census` for its Summary Job lineage and because it was more different from `nlcd_soils`): [remove-soil-texture.patch.txt](https://github.com/WikiWatershed/model-my-watershed/files/1314858/remove-soil-texture.patch.txt)

In the diff, I've changed the `nlcd_soil` task minimally, but if there are more efficient / Pythonic ways of working with it (extracting values into sublists), we should do that.

**Comment by rajadain** [comment]
Here's the differences in the GMS file for an identical watershed on this branch and staging:

```diff
--- SalleeCreekDeepCreek_HUC12_Stg.gms	2017-09-19 12:01:50.000000000 -0400
+++ SalleeCreekDeepCreek_HUC12_Dev.gms	2017-09-19 12:00:45.000000000 -0400
@@ -1,5 +1,5 @@
 10,6,0
-1.4.0,0.06,0,10,0,0,0.13157797203445618,14.039665441866896,50628,529.7894048397646,50098.210595160235,182.51873791206367,0,4.823605552111799,0.0025935746269670763,30,1961,1990,0.00028058108137844084,6241.366841944627,0,0,0,0.2919860497542336
+1.4.0,0.01,0,10,0,0,0.13157797203445618,14.039665441866896,50628,529.7894048397646,50098.210595160235,182.51873791206367,0,4.823605552111799,0.0025935746269670763,30,1961,1990,0.00028058108137844084,6241.366841944627,0,0,0,0.2919860497542336
 0
 0
 0
@@ -34,14 +34,14 @@
 Md_Residential,0.0,0.52,92,74,100
 Hd_Residential,0.0,0.87,92,74,110
 0,1,0,0,1,1,1.4
-723.3636363636364,501.71615452277615,0.3558582217973232,0.015339866156787764,0.25,0.25
+723.3636363636364,446.74545454545455,0.3558582217973232,0.015339866156787764,0.25,0.25
 2,1,1,1,1
-0.75,0.22010766845953078
-2.9,0.22010766845953078
+0.75,0.20464715909090908
+2.9,0.20464715909090908
 0.19,0.01
 0.19,0.01
 0.02,0.01
-2.5,1.087751830230301
+2.5,0.9685719334493256
 0.5,0.01
 0.3,0.01
 0.1,0.01
```

The differences are in: `RecessCoeff`, `Pload`, `PhosConc`

**Comment by rajadain** [comment]
This looks good and works well. I'll take another look once that patch has been added.

File: src/mmw/apps/modeling/mapshed/tasks.py [hunk 2]
```diff
@@ -244,41 +247,45 @@ def nlcd_streams_drb(result):
...
+        return {key: sum(counts(ng_keys))
+                for key, ng_keys in grouped.iteritems()}
 
-        # Map soil group values to usable subset
-        g2 = settings.SOIL_GROUP[g]
-        ng_count[(n, g2)] = count + ng_count.get((n, g2), 0)
+    # Group by land
+    n_count = count_by_group(lambda(n, g): n)
+    # Group by (land, usable soil group subset)
+    ng2_count = count_by_group(lambda(n, g): (n, settings.SOIL_GROUP[g]))
```
Comment (arottersman) [inline src/mmw/apps/modeling/mapshed/tasks.py]: I thought I was being clever, but this is actually slower than the for loop. About to push a fixup to put it back.

File: src/mmw/apps/modeling/mapshed/tasks.py [hunk 3]
```diff
@@ -152,6 +150,9 @@ def collect_data(geop_result, geojson):
 
     z['SedNitr'] = geop_result['soiln'] * 4.0
 
+    recess_coef = geop_result['recess_coef'] * -0.0015 + 0.1103
```
Comment (rajadain) [inline src/mmw/apps/modeling/mapshed/tasks.py]: Consider moving these calculations to within the `recess_coef` and `soilp` (and `soiln`) methods. Unless a value is assigned multiple times, we tend to not do any calculations here, instead doing everything within the "process" method after the geoprocessing. This helps keep these magic numbers close to original references, and they can be expounded on via comments without clutter.

**Comment by arottersman** [comment]
- Rebased on develop
- Added commit to move calculations out of `collect_data`
- Added commit to remove lint on develop

**Comment by rajadain** [comment]
Taking a final look.

**Comment by arottersman** [comment]
Thanks for the thorough review!

**Pull request merged**